src/gennylib/include/gennylib/parallel.hpp

```diff
@@ -0,0 +1,245 @@
+// Copyright 2019-present MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef HEADER_5129031F_B241_46DD_8285_64596CB0C155_INCLUDED
+#define HEADER_5129031F_B241_46DD_8285_64596CB0C155_INCLUDED
+
+#include <boost/throw_exception.hpp>
+#include <boost/exception/exception.hpp>
+#include <boost/log/trivial.hpp>
+
+#include <algorithm>
+#include <iterator>
+#include <variant>
+#include <thread>
+#include <mutex>
+#include <vector>
+#include <deque>
+
+namespace genny {
+
+/**
+ * Data structure that wraps STL containers and is thread-safe for insertions.
+ *
+ * Remain vigilant about the fact that this container offers no guarantees regarding
+ * the thread-safety of underlying container elements after returning.
+ *
+ * This container object also meets the BasicLockable requirement.
+ */
+template <typename T>
+class AtomicContainer {
+public:   
+    using value_type = typename T::value_type;
+
+
+    AtomicContainer() = default;
+    AtomicContainer(const AtomicContainer& rhs) {
+        std::lock_guard<lock_t> lock(rhs._mutex);
+        this->_container = rhs._container;
+    }
+    AtomicContainer(AtomicContainer&& rhs) {
+        std::lock_guard<lock_t> lock(rhs._mutex);
+        this->_container = std::move(rhs._container);
+    }
+
+    AtomicContainer<T>& operator=(const AtomicContainer& rhs) {
+        std::lock_guard<lock_t> lock(rhs._mutex);
+        this->_container = rhs._container;
+        return *this;
+    }
+    AtomicContainer<T>& operator=(AtomicContainer&& rhs) {
+        std::lock_guard<lock_t> lock(rhs._mutex);
+        this->_container = std::move(rhs._container);
+    }
```

Comment:
Do we gain anything by allowing copy/move construction or assignment? I think just inherit from boost::non_copyable and non_moveable or whatever they're called and we don't have to worry about the gotchas of these operations.

Reply:
Good point. I'd originally made it copyable/moveable since the ActorVector that we plan to use this for needs them. That said, we could probably just pointer-wrap that so disallowing those would be a bit easier.

Reply:
Ended up having to undo the the removal of copy/move since the `ParallelException` needed it to be safely throwable, and I preferred to lean on this code rather than implement a thread-safe insertable but throwable type there. Defs push back if you disagree.

That said, I think most of the correctness of these operators will depend on the underlying type; we just need to make sure lock shenanigans aren't happening.